WordPress/gutenberg
# Patterns: Add categories to user patterns, and allow filtering by these in site and post editor

## What?
Add the ability to assign categories to user created patterns.

To make reviewing easier separate PRs will be created off this branch for:

- Filtering by user categories in site editor - https://github.com/WordPress/gutenberg/pull/53837
- Filtering by user categories in post editor inserter

## Why?
Currently users have no way of organising their synced and unsynced patterns.

## How?
Uses the [recently added](https://github.com/WordPress/gutenberg/pull/53163) `wp_patterns_category` custom taxonomy, and adds a taxonomy input box to the create pattern modal.

## Testing Instructions
- In the post editor add a block and then in the block toolbar overflow select the `Create pattern` option
- Add categories for the pattern in the create pattern model and add the pattern
- Go to `/wp-admin/edit.php?post_type=wp_block` and check that the categories are attached to the new pattern
- Try adding another pattern and try a mix of adding existing and new category names and make sure they are all saved
- Also try adding a new pattern from the site editor Patterns page

### Site editor testing instructions
 - Add several user patterns, synced and unsynced, with categories assigned (make sure to hit enter or add ',' after each category - this is the default tag input component behaviour - we will fix this in a follow up PR)
 - Go to the site editor patterns page and make sure the categories appear in the left nav panel
 - Make sure that when the categories are selected the correct patterns display
 - Make sure the theme patterns also still display as expected
 - Make sure that all the patterns display under 'All patterns' and that searching works as expected
 - Duplicate a theme pattern and make sure the new pattern is in the same category
 - Add a user pattern with category of `Footer` and make sure it displays along with the theme Footer patterns (be aware that although the TT3 theme footer patterns are listed under `Footers` the category slug is actually `footer` so the user category name needs to match this).

### Post Editor testing
- In the post editor add some synced and unsynced patterns with categories assigned
- Check that the patterns appear in the correct categories in the inserter patterns tab
- Check that the `All patterns` tab shows all patterns and that paging at the bottom of the tab works
- Check that the Source Filter select list works as expected and that correct patterns display for each selected filter
- Also check that when the `My patterns` source filter is selected that a `Sync type filter` appears at the top of the patterns list and works as expected
- Check that changing Source or Sync type filters when categories are selected resets the category to `All`
- Repeat all of the above in the patterns explorer modal
- With screenreader/voiceover running try selecting a pattern source from dropdown at top of patterns tab and check that the number of category buttons displayed is announced correctly

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/3629020/d543a386-1b79-4496-84f9-7084642cdf2e
